tendrl/gluster_integration/sds_sync/__init__.py

```diff
@@ -930,3 +924,91 @@ def get_volume_alert_counts():
+            )
+            block_devices = {}
+            for dev_info in devlist:
+                if dev_info["TYPE"] == "disk":
+                    # create new block device list
+                    block_devices = {dev_info["KNAME"]: dev_info}
+                else:
+                    block_devices[dev_info["KNAME"]] = dev_info
+                # find ancestors if mountpoint present
+                if dev_info["MOUNTPOINT"]:
+                    ancestor_details = find_ancestors(
+                        block_devices, dev_info["KNAME"]
+                    )
+                    if dev_info["MOUNTPOINT"] in devicetree.keys():
+                        # if different block device mounted in same mountpoint
+                        # then append the disk and partition information under
+                        # same mountpoint
+                        disks = devicetree[dev_info["MOUNTPOINT"]]["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = list(
+                            set(disks + ancestor_details["disks"])
+                        )
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            list(
+                                set(ancestor_details["partitions"])
+                            )
+                    else:
+                        devicetree[dev_info["MOUNTPOINT"]] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["disks"] = \
+                            ancestor_details["disks"]
+                        devicetree[dev_info["MOUNTPOINT"]]["partitions"] = \
+                            ancestor_details["partitions"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"] = {}
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "mountpoint"] = dev_info["MOUNTPOINT"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "type"] = dev_info["TYPE"]
+                        devicetree[dev_info["MOUNTPOINT"]]["device_info"][
+                            "size"] = dev_info["SIZE"]
+        return devicetree
+    except Exception as ex:
+        logger.log(
+            "error",
+            NS.publisher_id,
+            {"message": "Unable to generate devicetree.err: %s" % ex}
+        )
+
+
+def find_ancestors(block_devices, device_name):
+    # find ancestors including current dev from bottom to top
+    ancestor_details = {}
+    ancestor_details["disks"] = []
+    ancestor_details["partitions"] = []
+    while True:
+        if device_name in block_devices.keys():
+            if block_devices[device_name]["TYPE"] == "disk":
+                ancestor_details["disks"].append(
+                    block_devices[device_name]["NAME"]
+                )
+                break
```

Comment:
First device found, we break. Right?

Reply:
yes,
let's take example lsblk output like : 

vdb                                 252:16   0  256M  0 disk 
├─GLUSTER_vg1-GLUSTER_pool1_tmeta   253:2    0    4M  0 lvm  
│ └─GLUSTER_vg1-GLUSTER_pool1-tpool 253:7    0  248M  0 lvm  
│   ├─GLUSTER_vg1-GLUSTER_pool1     253:9    0  248M  0 lvm  
│   └─GLUSTER_vg1-GLUSTER_lv1       253:11   0  248M  0 lvm  /gluster/brick1
└─GLUSTER_vg1-GLUSTER_pool1_tdata   253:3    0  248M  0 lvm  
  └─GLUSTER_vg1-GLUSTER_pool1-tpool 253:7    0  248M  0 lvm  
    ├─GLUSTER_vg1-GLUSTER_pool1     253:9    0  248M  0 lvm  
    └─GLUSTER_vg1-GLUSTER_lv1       253:11   0  248M  0 lvm  /gluster/brick1

this function is called when the mount point is found so here mount point is  /gluster/brick1, then it will loop using its parent name "GLUSTER_vg1-GLUSTER_pool1-tpool" (it will loop using kernel name of this), this is also not "disk" so again loop using its parent name "GLUSTER_vg1-GLUSTER_pool1_tmeta"  this is also not disk again loop using its parent name "vdb" now we found disk then it will break;


if you see here there are two "/gluster/brick1", but both will give the same disk information. second will override information which we collected earlier because both are the same mount point. overriding previously collected info not a problem I feel.

Reply:
Ack. Can you explain the same using comments in code?